In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns 
from scipy.stats import pointbiserialr

# Bags to bags analysis 

In [2]:
bags_stats = pd.read_csv('bag_statistics.csv')

very = bags_stats.sort_values(by='Superlat', ascending=False)
names = bags_stats.sort_values(by='Names', ascending=False)
me_you = bags_stats.sort_values(by='Me/You', ascending=False)
no_yes = bags_stats.sort_values(by='No/Yes', ascending=False)
I_E = bags_stats.sort_values(by='Bags I/E', ascending=False)
N_S = bags_stats.sort_values(by='Bags N/S', ascending=False)
F_T = bags_stats.sort_values(by='Bags F/T', ascending=False)
P_J = bags_stats.sort_values(by='Bags P/J', ascending=False)
#I_E

speaker  MBTI  F1  F2                                      words_counted  \
15       Roy  ESFP  Se  Fi  {'hey': 30, 'man': 12, 'baby': 2, 'no': 17, 'c...   
12     Oscar  INTJ  Ni  Te  {'you': 184, 'guys': 6, 'better': 3, 'update':...   
17   Stanley  ISTJ  Si  Te  {'well': 8, 'it': 39, 'could': 6, 'be': 23, 'o...   
14   Phyllis  ISFJ  Si  Fe  {'so': 28, 'what': 34, 'does': 8, 'downsizing'...   
7      Karen  ISTJ  Si  Te  {'jims': 6, 'nice': 5, 'enough': 3, 'i': 85, '...   
6        Jim  ENTP  Ne  Ti  {'oh': 326, 'i': 1088, 'told': 23, 'you': 1142...   
9      Kevin  ISFP  Fi  Se  {'yeah': 47, 'itll': 2, 'be': 28, 'you': 179, ...   
13       Pam  ISFJ  Si  Fe  {'well': 69, 'i': 1010, 'dont': 176, 'know': 1...   
8      Kelly  ESFP  Se  Fi  {'i': 288, 'have': 41, 'a': 118, 'customer': 1...   
2     Darryl  ISTP  Ti  Se  {'its': 26, 'not': 23, 'my': 21, 'real': 2, 'n...   
18      Toby  ISFJ  Si  Fe  {'hey': 44, 'were': 7, 'not': 38, 'all': 13, '...   
4      Holly  INFP  Fi  Ne  {'hi': 3, 'yes': 4, 'uh': 4, 'im': 23, 'holly'...   
11   Michael  ENFP  Ne  Fi  {'all': 723, 'right': 605, 'jim': 232, 'your':...   
0       Andy  ESFP  Se  Fi  {'hey': 33, 'big': 25, 'tuna': 50, 'youre': 48...   
5        Jan  ENTJ  Te  Ni  {'on': 89, 'her': 9, 'cell': 4, 'phone': 24, '...   
3     Dwight  ESTJ  Te  Si  {'singing': 9, 'shall': 5, 'i': 1341, 'play': ...   
16      Ryan  ESTP  Se  Ti  {'hey': 32, 'ryan': 20, 'howard': 3, 'from': 1...   
1     Angela  ISTJ  Si  Te  {'i': 323, 'bet': 1, 'its': 59, 'gon': 8, 'na'...   
10  Meredith  ESTP  Se  Ti  {'michael': 10, 'please': 5, 'can': 9, 'we': 1...   

    Nb Introv  Nb Extrav  Superlat     Names     No/Yes     Me/You   Bags I/E  \
15         13          9  1.455399  2.018779  57.861635  18.300654  59.090909   
12         35         34  1.092683  1.697561  47.413793  22.261484  50.724638   
17         22         22  0.791701  1.283101  68.320611  36.842105  50.000000   
14         29         36  0.940893  2.243667  55.677656  25.714286  44.615385   
7          23         29  1.442469  2.314660  40.096618  29.411765  44.230769   
6         172        222  1.434514  2.052364  48.411142  22.490185  43.654822   
9          32         42  1.141062  2.530180  53.906250  31.629393  43.243243   
13        112        151  1.407434  2.333694  52.025932  29.583975  42.585551   
8          29         41  1.283106  2.204310  57.432432  34.375000  41.428571   
2          12         17  0.505732  0.741740  62.500000  20.000000  41.379310   
18         21         30  1.398058  1.825243  50.712251  22.058824  41.176471   
4          13         19  1.434570  1.154654  52.674897  24.637681  40.625000   
11        505        859  1.029569  1.893452  49.017353  27.399651  37.023460   
0          49         90  1.145096  1.340967  48.258706  33.668342  35.251799   
5          44         84  1.166937  0.929227  51.091703  23.501577  34.375000   
3         106        208  0.885321  1.511468  53.140265  29.886982  33.757962   
16         29         66  1.275111  1.720197  51.570248  28.099174  30.526316   
1          38         87  0.665161  2.171185  63.255034  22.888889  30.400000   
10          4         10  1.183432  1.563821  65.365854  40.845070  28.571429   

     Bags N/S   Bags F/T   Bags P/J  
15  25.000000  92.307692  65.384615  
12  46.666667  69.696970  68.181818  
17  53.333333  73.076923  67.441860  
14  57.142857  78.947368  67.241379  
7   44.444444  76.190476  65.853659  
6   62.500000  78.321678  64.069264  
9   40.000000  77.272727  75.757576  
13  42.000000  87.500000  68.896321  
8   57.142857  88.095238  74.647887  
2   40.000000  92.857143  71.428571  
18  30.769231  56.000000  73.333333  
4   71.428571  72.727273  80.434783  
11  62.458472  81.614350  75.324675  
0   33.333333  78.571429  68.148148  
5   48.484848  75.000000  70.909091  
3   40.298507  75.598086  69.696970  
16  50.000000  70.454545  69.000000  
1   50.000000  75.862069  59.770115  
10  60.000000  72.727273  57.142857

In [3]:
def stats_encoder(df):
    df['I'] = df['MBTI'].apply(lambda x: 'I' in x)
    df['N'] = df['MBTI'].apply(lambda x: 'N' in x)
    df['F'] = df['MBTI'].apply(lambda x: 'F' in x)
    df['P'] = df['MBTI'].apply(lambda x: 'P' in x)

    df['Te'] = df.apply(lambda x: 'Te' in x['F1'] or 'Te' in x['F2'], axis=1)
    df['Ti'] = df.apply(lambda x: 'Ti' in x['F1'] or 'Ti' in x['F2'], axis=1)
    df['Se'] = df.apply(lambda x: 'Se' in x['F1'] or 'Se' in x['F2'], axis=1)
    df['Si'] = df.apply(lambda x: 'Si' in x['F1'] or 'Si' in x['F2'], axis=1)
    df['Fe'] = df.apply(lambda x: 'Fe' in x['F1'] or 'Fe' in x['F2'], axis=1)
    df['Fi'] = df.apply(lambda x: 'Fi' in x['F1'] or 'Fi' in x['F2'], axis=1)
    df['Ne'] = df.apply(lambda x: 'Ne' in x['F1'] or 'Ne' in x['F2'], axis=1)
    df['Ni'] = df.apply(lambda x: 'Ni' in x['F1'] or 'Ni' in x['F2'], axis=1)
    
    return df

In [4]:
bags_stats = stats_encoder(bags_stats)

correlations = {}

# Iterate over boolean columns
for bool_col in ['I', 'N','F', 'P', 'Ne', 'Ni', 'Fe', 'Fi', 'Se', 'Si', 'Te', 'Ti']:
    # Iterate over numerical columns
    for num_col in ['Bags I/E','Bags N/S','Bags F/T','Bags P/J', 'Superlat', 'Names','No/Yes', 'Me/You']:
        # Calculate correlation coefficient and p-value
        correlation_coefficient, p_value = pointbiserialr(bags_stats[bool_col], bags_stats[num_col])
        
        # Store results in dictionary
        correlations[f'{bool_col} vs {num_col}'] = {
            'correlation_coefficient': correlation_coefficient,
            'p_value': p_value
        }

for key, value in correlations.items():
    if (abs(value['correlation_coefficient']) > 0.3) and (value['p_value'] < 0.4):
        print(f"{key}: correlation_coefficient={value['correlation_coefficient']}, p_value={value['p_value']}")

I vs Bags I/E: correlation_coefficient=0.3147890587324854, p_value=0.18930030845261717
N vs Bags N/S: correlation_coefficient=0.5190827452421374, p_value=0.02276044135366906
N vs No/Yes: correlation_coefficient=-0.3886273736586746, p_value=0.100100293598989
N vs Me/You: correlation_coefficient=-0.3566317345975871, p_value=0.1339306090992606
F vs Bags P/J: correlation_coefficient=0.5330754351920061, p_value=0.0187647079456743
F vs Superlat: correlation_coefficient=0.3737359668589323, p_value=0.11497359023025826
F vs Names: correlation_coefficient=0.3558286760999603, p_value=0.13487143427626397
P vs Bags F/T: correlation_coefficient=0.3737346146207633, p_value=0.11497500860881514
Ne vs Bags N/S: correlation_coefficient=0.641237360536208, p_value=0.0030879689890344606
Ne vs Bags P/J: correlation_coefficient=0.3349979412743628, p_value=0.16091854255367652
Ni vs Names: correlation_coefficient=-0.3144075582384394, p_value=0.18986630455155828
Fe vs Names: correlation_coefficient=0.32489199540

### Bag of words -> MBTI observations

- Bags N/S -> N : correlation_coefficient=0.5190827452421374, p_value=0.02276044135366906
- Bags N/S -> Ne : correlation_coefficient=0.641237360536208, p_value=0.0030879689890344606

- Bags I/E -> I : correlation_coefficient=0.3147890587324854, p_value=0.18930030845261717

# Sentiment analysis

## 1. Vader

In [5]:
vader = pd.read_csv('vader_scores.csv')
pos = vader.sort_values(by='vader_pos', ascending=False)
neg = vader.sort_values(by='vader_neg', ascending=False)
neu = vader.sort_values(by='vader_neu', ascending=False)
comp = vader.sort_values(by='vader_comp', ascending=False)

In [11]:
df = stats_encoder(vader)
#df

In [7]:
correlations = {}

# Iterate over boolean columns
for bool_col in ['I', 'N','F', 'P', 'Ne', 'Ni', 'Fe', 'Fi', 'Se', 'Si', 'Te', 'Ti']:
    # Iterate over numerical columns
    for num_col in ['vader_pos', 'vader_neg', 'vader_neu', 'vader_comp','vader_power']:
        # Calculate correlation coefficient and p-value
        correlation_coefficient, p_value = pointbiserialr(vader[bool_col], vader[num_col])
        
        # Store results in dictionary
        correlations[f'{bool_col} vs {num_col}'] = {
            'correlation_coefficient': correlation_coefficient,
            'p_value': p_value
        }

for key, value in correlations.items():
    if (abs(value['correlation_coefficient']) > 0.3) and (value['p_value'] < 0.4):
        print(f"{key}: correlation_coefficient={value['correlation_coefficient']}, p_value={value['p_value']}")

N vs vader_pos: correlation_coefficient=0.39356034569084697, p_value=0.09549559115842142
Ne vs vader_pos: correlation_coefficient=0.40046392990280766, p_value=0.08931266938293489
Ne vs vader_neu: correlation_coefficient=-0.38806623663302475, p_value=0.10063408880283738
Ne vs vader_power: correlation_coefficient=0.3879703405713001, p_value=0.10072551791954355
Ti vs vader_comp: correlation_coefficient=-0.4069567963361717, p_value=0.08376960825288615


## 2. TextBlob

In [12]:
textblob = pd.read_csv('textblob.csv')
polarity = textblob.sort_values(by='polarity', ascending=False)
subj = textblob.sort_values(by='subjectivity', ascending=False)
#polarity

In [9]:
textblob = stats_encoder(textblob)

In [10]:
correlations = {}

# Iterate over boolean columns
for bool_col in ['I', 'N','F', 'P', 'Ne', 'Ni', 'Fe', 'Fi', 'Se', 'Si', 'Te', 'Ti']:
    # Iterate over numerical columns
    for num_col in ['polarity', 'subjectivity']:
        # Calculate correlation coefficient and p-value
        correlation_coefficient, p_value = pointbiserialr(textblob[bool_col], textblob[num_col])
        
        # Store results in dictionary
        correlations[f'{bool_col} vs {num_col}'] = {
            'correlation_coefficient': correlation_coefficient,
            'p_value': p_value
        }

for key, value in correlations.items():
    if (abs(value['correlation_coefficient']) > 0.3) and (value['p_value'] < 0.4):
        print(f"{key}: correlation_coefficient={value['correlation_coefficient']}, p_value={value['p_value']}")

I vs subjectivity: correlation_coefficient=-0.42214211249103767, p_value=0.0717981724242417
N vs polarity: correlation_coefficient=0.49208850897903555, p_value=0.032342212722514994
Ne vs polarity: correlation_coefficient=0.34230065734190596, p_value=0.1514230640920369
Fi vs subjectivity: correlation_coefficient=0.30422017329030704, p_value=0.2054005334136113
